pkg/distributors/contracts/MultiRewards.sol

```diff
@@ -70,191 +75,294 @@ contract MultiRewards is IMultiRewards, IDistributor, ReentrancyGuard, Temporari
+        IERC20 rewardsToken
+    ) external view returns (uint256) {
         return
-            rewardData[stakingToken][rewardsToken].rewardRate.mulDown(
-                rewardData[stakingToken][rewardsToken].rewardsDuration
+            rewardData[pool][rewarder][rewardsToken].rewardRate.mulDown(
+                rewardData[pool][rewarder][rewardsToken].rewardsDuration
             );
     }
 
     /* ========== MUTATIVE FUNCTIONS ========== */
-
-    function setRewardsDistributor(
-        IERC20 stakingToken,
-        IERC20 rewardsToken,
-        address rewardsDistributor
-    ) external onlyOwner {
-        rewardData[stakingToken][rewardsToken].rewardsDistributor = rewardsDistributor;
-    }
-
-    function stake(IERC20 stakingToken, uint256 amount) external {
-        stake(stakingToken, amount, msg.sender);
+    function stake(IERC20 pool, uint256 amount) external {
+        stake(pool, amount, msg.sender);
     }
 
+    /**
+     * @notice Stakes a token so that `receiver` can earn rewards
+     * @param pool The token being staked to earn rewards
+     * @param amount Amount of `pool` to stake
+     * @param receiver The recipient of claimed rewards
+     */
     function stake(
-        IERC20 stakingToken,
+        IERC20 pool,
         uint256 amount,
         address receiver
-    ) public nonReentrant whenNotPaused updateReward(stakingToken, receiver) {
+    ) public nonReentrant whenNotPaused updateReward(pool, receiver) {
         require(amount > 0, "Cannot stake 0");
-        _totalSupply[stakingToken] = _totalSupply[stakingToken].add(amount);
-        _balances[stakingToken][receiver] = _balances[stakingToken][receiver].add(amount);
-        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
-        emit Staked(address(stakingToken), receiver, amount);
+        _totalSupply[pool] = _totalSupply[pool].add(amount);
+        _balances[pool][receiver] = _balances[pool][receiver].add(amount);
+        pool.safeTransferFrom(msg.sender, address(this), amount);
+        emit Staked(address(pool), receiver, amount);
     }
 
     /**
      * @notice Stake tokens using a permit signature for approval
+     * @param pool The bpt being staked to earn rewards
      * @param amount    Amount of allowance
      * @param deadline  The time at which this expires (unix time)
      * @param v         v of the signature
      * @param r         r of the signature
      * @param s         s of the signature
      */
     function stakeWithPermit(
-        IERC20 stakingToken,
+        IERC20 pool,
         uint256 amount,
         uint256 deadline,
         uint8 v,
         bytes32 r,
         bytes32 s
     ) public {
-        IERC20Permit(address(stakingToken)).permit(msg.sender, address(this), amount, deadline, v, r, s);
-        stake(stakingToken, amount, msg.sender);
+        IERC20Permit(address(pool)).permit(msg.sender, address(this), amount, deadline, v, r, s);
+        stake(pool, amount, msg.sender);
```

Comment:
If we care about arbitrary recipients, shouldn't we allow for one to be passed here?

Reply:
A user `A` who stakes their tokens to a recipient `B` would also need to be able to unstake them back.  So this would require another level of accounting right?  I think it's better to keep stake as for the benefit of `msg.sender`.  Is there a use case you are imagining?

Reply:
Would like to address this as part of https://github.com/balancer-labs/balancer-v2-monorepo/issues/659